### PR TITLE
[Ruby client] add validation to nil parameter

### PR DESCRIPTION
**Pull request by linr211**
Changes of this PR:
1. Add validation to nil locale parameter in the API
2. fix issue https://github.com/vmware/singleton/issues/1479
3. fix issue https://github.com/vmware/singleton/issues/1484
4. Change default_language to source_locale
5. Change to log level to debug when the key is missed

### CODE REVIEW

**Review by jessiejuachon — DISMISSED** [review]

File: lib/sgtn-client/api/source.rb [hunk 1]
```diff
@@ -40,6 +40,10 @@ def self.getSources(component, locale)
 
       def self.loadBundles(locale)
         SgtnClient.logger.debug "[Source][loadBundles]locale=#{locale}"
+        if locale.nil?
+          SgtnClient.logger.error "[Source][loadBundles]the input is nil!"
```
[the remark this answers is not in the document]
Comment (linr211) [reply lib/sgtn-client/api/source.rb]: fixed.

**Comment by Xiaochao8** [comment]
I wonder what's the reason to 'Change default_language to source_locale'?

**Comment by linr211** [comment]
> I wonder what's the reason to 'Change default_language to source_locale'?

Jessie mentioned it in a email thread:
1. We should use “source” instead of “default” because we actually mean the source or origin of translation.
2. We should use “locale” instead of “language”. Technically, a locale (or language tag) can be composed of many subtags: language-extlang-script-region-variant-extension-privateuse. At the minimum, a locale may contain only the language subtag (e.g. “en”). In our case,  the source locale is actually en-US (i.e. language-region).

File: lib/sgtn-client/api/translation.rb [hunk 2]
```diff
@@ -11,25 +11,25 @@ module Core
...
-      def self.getString(component, key, locale)
-        SgtnClient.logger.debug "[Translation][getString_p]component=#{component}, key=#{key}, locale=#{locale}"
+      def self.getString(component, key, locale=SgtnClient::Config.configurations.default)
+        SgtnClient.logger.debug "[Translation][getString]component=#{component}, key=#{key}, locale=#{locale}"
         str = getTranslation(component, key, locale)
         if str.nil?
           str = SgtnClient::Source.getSource(component, key, SgtnClient::Config.configurations.default)
           if str.nil?
-            SgtnClient.logger.error "[Translation]can't find the key '" + key + "' in source path!"
+            SgtnClient.logger.debug "[Translation][getString]can't find key=#{key}, component=#{component}, locale=#{locale}!!!"
```
Comment (jessiejuachon) [inline lib/sgtn-client/api/translation.rb:20]: This is for a different, higher-priority issue: https://github.com/vmware/singleton/issues/1493
Next time, please keep the PR as small as possible.

File: lib/sgtn-client/api/translation.rb [hunk 3]
```diff
@@ -11,25 +11,25 @@ module Core
...
 
-      def self.getString_p(component, key, plural_args, locale)
+      def self.getString_p(component, key, plural_args, locale=SgtnClient::Config.configurations.default)
         SgtnClient.logger.debug "[Translation][getString_p]component=#{component}, key=#{key}, locale=#{locale}"
         str = getTranslation(component, key, locale)
         if str.nil?
           str = SgtnClient::Source.getSource(component, key, SgtnClient::Config.configurations.default)
           if str.nil?
-            SgtnClient.logger.error "[Translation]can't find the key '" + key + "' in source path!"
+            SgtnClient.logger.debug "[Translation][getString_p]can't find key=#{key}, component=#{component}, locale=#{locale}!!!"
```
Comment (jessiejuachon) [inline lib/sgtn-client/api/translation.rb:32]: This is for a different, higher-priority issue: https://github.com/vmware/singleton/issues/1493
Next time, please keep the PR as small as possible.

File: lib/sgtn-client/api/translation.rb [hunk 4]
```diff
@@ -54,15 +54,15 @@ def self.getString_f(component, key, args, locale, *optionals)
...
+      def self.getStrings(component, locale=SgtnClient::Config.configurations.default)
         SgtnClient.logger.debug "[Translation][getStrings]component=#{component}, locale=#{locale}"
         items = get_cs(component, locale)
         default = SgtnClient::Config.configurations.default
         if items.nil? || items["messages"] == nil
           items = {}
           s = SgtnClient::Source.getSources(component, default)
           if s.nil?
-            SgtnClient.logger.error "[Translation]can't find the component '" + component + "' in source path!"
+            SgtnClient.logger.debug "[Translation][getStrings]can't find component=#{component}, locale=#{locale}!!!"
```
Comment (jessiejuachon) [inline lib/sgtn-client/api/translation.rb:65]: This is for a different, higher-priority issue: https://github.com/vmware/singleton/issues/1493
Next time, please keep the PR as small as possible.

File: lib/sgtn-client/util/locale-util.rb [hunk 5]
```diff
@@ -11,14 +11,17 @@ module SgtnClient
...
+            end
             found = SgtnClient::DEFAULT_LOCALES.select {|e| e == locale}
             if !found.empty?
                 return found[0]
             end
             if SgtnClient::MAP_LOCALES.key?(locale)
                 return SgtnClient::MAP_LOCALES[locale]
             end
-            parts = locale.split("-")
+            parts = locale.to_s.split("-")
```
Comment (jessiejuachon) [inline lib/sgtn-client/util/locale-util.rb:24]: This is for a different issue: https://github.com/vmware/singleton/issues/1479
Please move to a different PR.

File: lib/sgtn-client/util/cache-util.rb [hunk 6]
```diff
@@ -35,16 +38,19 @@ def self.write_cache(cache_key, items)
...
         version = SgtnClient::Config.configurations[env]["version"].to_s
-        default_l = SgtnClient::Config.configurations[env]["default_language"]
+        default_l = SgtnClient::Config.configurations[env]["source_locale"]
         if default_l == nil
           default_l = 'en'
         end
         lc = locale == default_l ? SgtnClient::Config.configurations.default: locale
         SgtnClient.logger.debug "[CacheUtil]get cache key: #{lc}"
-        return product_name + "_" + version + "_" + component + "_" + lc
+        return product_name + "_" + version + "_" + component + "_" + lc.to_s
```
Comment (jessiejuachon) [inline lib/sgtn-client/util/cache-util.rb:53]: This is for a different issue: https://github.com/vmware/singleton/issues/1479
Please move to a different PR.

**Review by jessiejuachon — CHANGES_REQUESTED** [review]

**Comment by jessiejuachon** [comment]
I realize some code that are unrelated might cause an issue separately. Also, some code fixes are higher priority, so requesting to break this into smaller PRs and link to appropriate issues logged.

**Review by jessiejuachon — CHANGES_REQUESTED** [review]

File: lib/sgtn-client/api/source.rb [hunk 7]
```diff
@@ -40,6 +40,10 @@ def self.getSources(component, locale)
 
       def self.loadBundles(locale)
         SgtnClient.logger.debug "[Source][loadBundles]locale=#{locale}"
+        if locale.nil?
```
Comment (jessiejuachon) [inline lib/sgtn-client/api/source.rb:43]: Do not hide the exception here as nil calue for locale is invalid for this API method.

**Comment by linr211** [comment]
> 

sure

**Pull request closed without merging**